onnxruntime/core/optimizer/graph_transformer_mgr.cc

```diff
@@ -2,24 +2,48 @@
 // Licensed under the MIT License.
 
 #include "core/optimizer/graph_transformer_mgr.h"
+#include "core/optimizer/rule_based_graph_transformer.h"
 using namespace onnxruntime;
 using namespace ::onnxruntime::common;
 
 namespace onnxruntime {
 
-Status GraphTransformerManager::ApplyAll(Graph& graph) const {
+common::Status GraphTransformerManager::ApplyTransformers(Graph& graph, std::vector<std::string> providers, const TransformerLevel& level) const {
+  const auto& transformers = level_to_transformer_map_.find(level);
+  if (transformers == level_to_transformer_map_.end()) {
+    return Status::OK();
+  }
+
+  // sort the providers. This makes it easy to find intersection of requested providers for this run
+  // and registered providers for a transformer
+  std::sort(providers.begin(), providers.end());
+
   for (unsigned step = 0; step < steps_; ++step) {
-    bool changed = false;
-    for (auto& transformer : transformers_) {
-      bool t_changed = false;
-      Status s = transformer->Apply(graph, t_changed);
-      if (!s.IsOK()) {
-        return s;
-      }
-      changed = changed || t_changed;
+    bool graph_changed = false;
+    for (const auto& transformer : transformers->second) {
+      const auto compatible_providers = GenerateCompatibleProvidersList(providers, transformer->Name());
+      bool modified = false;
```

Comment:
I don't quite understand this. I expected you just care that the transformer matches a name in the providers list, and not which one/s it matches. This code seems to get a list of all matches and calls transformer->Apply regardless of whether that list is empty or not. However the list of providers is ignored in graph_transformer.cc anyway, so not sure what purpose it serves.

Reply:
here - I need to add a check to call apply only if this list is not empty... I have it in my next update

In graph_transformers.cc this list is ignored right now because Right now none of the transformers check providers... I need to make that change for all the transformers ... + we may be converting them to rewrite rules ... Therefore I am going to leave it out of this PR...

When I check in this PR I am not going to create any transformers other than EliminateSlice and EliminateIdentity... IN the next PR when I make necessary fixes I will enable them

---
In reply to: [265383041](https://github.com/Microsoft/onnxruntime/pull/608#discussion_r265383041) [](ancestors = 265383041)

Reply:
I am simplifying this more... I am removing providers during apply.... right now we only have 2 phases before and after partition...after partition we dont need to choose only a couple of providers to apply... so we can make this simple and simply run all transformers registered for the level...

---
In reply to: [265384224](https://github.com/Microsoft/onnxruntime/pull/608#discussion_r265384224) [](ancestors = 265384224,265383041)